python/ql/test/experimental/dataflow/coverage/test.py

```diff
@@ -39,93 +42,241 @@ def test_floatnumber_literal():
     x = {"s": SOURCE}
-    SINK(x["s"])
+    SINK(x["s"]) # Flow missing
 
 def test_dict_comprehension():
     x = {y: SOURCE for y in ["s"]}
-    SINK(x["s"])
+    SINK(x["s"]) # Flow missing
 
 def test_nested_dict_display():
     x = {** {"s": SOURCE}}
-    SINK(x["s"])
+    SINK(x["s"]) # Flow missing
 
 # 6.2.8. Generator expressions
 def test_generator():
     x = (SOURCE for y in [3])
-    SINK([*x][0])
+    SINK([*x][0]) # Flow missing
 
-# List taken from https://docs.python.org/3/reference/expressions.html
-# 6. Expressions
-# 6.1. Arithmetic conversions
-# 6.2. Atoms
-# 6.2.1. Identifiers (Names)
-# 6.2.2. Literals
-# 6.2.3. Parenthesized forms
-# 6.2.4. Displays for lists, sets and dictionaries
-# 6.2.5. List displays
-# 6.2.6. Set displays
-# 6.2.7. Dictionary displays
-# 6.2.8. Generator expressions
 # 6.2.9. Yield expressions
+def gen(x):
+    yield x
+
+def test_yield():
+    g = gen(SOURCE)
+    SINK(next(g)) # Flow missing
+
+def gen_from(x):
+    yield from gen(x)
+
+def test_yield_from():
+    g = gen_from(SOURCE)
+    SINK(next(g)) # Flow missing
+
+# a statement rather than an expression, but related to generators
+def test_for():
+    for x in gen(SOURCE):
+        SINK(x) # Flow missing
+
 # 6.2.9.1. Generator-iterator methods
-# 6.2.9.2. Examples
+def test___next__():
+    g = gen(SOURCE)
+    SINK(g.__next__()) # Flow missing
+
+def gen2(x):
+    m = yield x # argument of `send` has to flow to value of `yield x` (and so to `m`)
+    yield m
+
+def test_send():
+    g = gen2(3)
+    n = next(g)
+    SINK(g.send(SOURCE)) # Flow missing
+
+def gen_ex(x):
+    try:
+        yield 3
+    except:
+        yield x # `x` has to flow to call to `throw`
+
+def test_throw():
+    g = gen_ex(SOURCE)
+    n = next(g)
+    SINK(g.throw(TypeError)) # Flow missing
+
+# no `test_close` as `close` involves no data flow
+
 # 6.2.9.3. Asynchronous generator functions
+async def agen(x):
+    yield x
+
 # 6.2.9.4. Asynchronous generator-iterator methods
-# 6.3. Primaries
+
+# helper to run async test functions
+def runa(a):
+    import asyncio
+    asyncio.run(a)
+
+async def atest___anext__():
+    g = agen(SOURCE)
+    SINK(await g.__anext__()) # Flow missing
+
+def test___anext__():
+    runa(atest___anext__())
+
+async def agen2(x):
+    m = yield x # argument of `send` has to flow to value of `yield x` (and so to `m`)
+    yield m
+
+async def atest_asend():
+    g = agen2(3)
+    n = await g.__anext__()
+    SINK(await g.asend(SOURCE)) # Flow missing
+
+def test_asend():
+    runa(atest_asend())
+
+async def agen_ex(x):
+    try:
+        yield 3
+    except:
+        yield x # `x` has to flow to call to `athrow`
+
+async def atest_athrow():
+    g = agen_ex(SOURCE)
+    n = await g.__anext__()
+    SINK(await g.athrow(TypeError)) # Flow missing
+
+def test_athrow():
+    runa(atest_athrow())
+
 # 6.3.1. Attribute references
+class C:
+    a = SOURCE
+
+def test_attribute_reference():
+    SINK(C.a) # Flow missing
```

Comment:
Could also test the double-nesting of attributes of attributes like `C.a.b` etc.

These seem to be static variables (sorry I don't read Python so well). How about some field-flow tests using instance variables, checking whether data can "leak" between different instances of the same class (which they shouldn't).

Reply:
Yes, double-nesting should probably be added. One could hope that if we have `.x` then we have it of all relevant things, including `a.b`, but we should perhaps not assume that..

Reply:
I have not added any instances for now (and hence no double-nesting).